heuristic/alns.py

```diff
@@ -0,0 +1,134 @@
+
+from heuristic.utils import WeightUpdate
+
+
+class ALNS:
+    def __init__(self, state, criterion):
+
+        self.initial_solution = state
+        self.current_solution = state
+        self.best_solution = state
+
+        self.random_state = self.initialize_random_state()
+        self.criterion = criterion
+
+        self.destroy_operators = {}
+        self.destroy_weights = {}
+        self.repair_operators = {}
+        self.repair_weights = {}
+
+    def iterate(self, iterations):
+        """
+        Performs the given number of iterations of ALNS by successively destroy and then repair the current solution
+
+        :param iterations: the number of iterations
+        :return: the best solution found
+        """
+
+        # todo: this initialization can be moved to a better place, but need an integrated method for setting operators
+        # before this method is called. If the operators are not initialized this method will raise DivisionByZeroError
+        self.initialize_destroy_and_repair_weights()
+
+        for iteration in range(iterations):
+
+            destroy_operator, destroy_id = self.select_operator(self.destroy_operators, self.destroy_weights)
+            repair_operator, repair_id = self.select_operator(self.repair_operators, self.repair_weights)
+
+            destroyed_solution = destroy_operator(self.current_solution)
+            candidate_solution = repair_operator(destroyed_solution)
+
+            self.consider_candidate_and_update_weights(candidate_solution, destroy_id, repair_id)
+
+        return self.best_solution
+
+    def consider_candidate_and_update_weights(self, candidate_solution, destroy_id, repair_id):
+        """
+        Considers the candidate based on self.critertion, and will update the weights according to the outcome
+
+        :param candidate_solution: The solution to consider
+        :param destroy_id: the id (name) of the destroy function used to create this state
+        :param repair_id: the id (name) of the repair function used to create this state
+        """
+
+        # todo: this has potential for performance improvements, but unsure if it is only for GreedyCriterion
+
+        if self.criterion.accept(candidate_solution, self.current_solution):
+            self.current_solution = candidate_solution
+
+            if candidate_solution.get_objectice_value() >= self.current_solution.get_objectice_value():
+                weight_update = WeightUpdate.IS_BETTER
+            else:
+                weight_update = WeightUpdate.IS_ACCEPTED
+        else:
+            weight_update = WeightUpdate.IS_REJECTED
+
+        if candidate_solution.get_objectice_value() >= self.best_solution.get_objective_value():
+
+            # todo: this is copied from the Github-repo, but unsure if this is the correct way to do it.
+            weight_update = WeightUpdate.IS_BEST
+            self.best_solution = candidate_solution
+            self.current_solution = candidate_solution
```

Comment:
In ALNS, this is one the uglier parts, so I understand completely why you're unsure. Looks to me you have a maximisation problem, whereas my code of course assumes minimisation. That's easily circumvented by crafting a dummy `State` class as follows:
```python
class State:

  def objective(self):
    # for the ALNS framework
    return -self.get_objective_value()
  
  def get_objective_value():
    return <your computations>
``` 

Good work guys! I translated your initial comment @ekallevik - if there's anything ALNS does not yet do that you need, would you care to let me know? I am definitely intent on improving my package, so it covers more use cases. If something's unclear, feel free to open an issue - I eventually want to add a readthedocs, but I have so far not yet come around to it.